src/awkward/contents/content.py

```diff
@@ -762,41 +761,23 @@ def _merging_strategy(
 
         head = [self]
         tail = []
-        i = 0
-
-        while i < len(others):
-            other = others[i]
-            if isinstance(
-                other,
-                (
-                    ak.contents.IndexedArray,
-                    ak.contents.IndexedOptionArray,
-                    ak.contents.ByteMaskedArray,
-                    ak.contents.BitMaskedArray,
-                    ak.contents.UnmaskedArray,
-                    ak.contents.UnionArray,
-                ),
-            ):
+
+        it_others = iter(others)
+
+        for other in it_others:
+            if other.is_indexed or other.is_option or other.is_union:
+                tail.append(other)
+                tail.extend(it_others)
                 break
             else:
                 head.append(other)
-            i = i + 1
-
-        while i < len(others):
-            tail.append(others[i])
-            i = i + 1
-
-        if any(isinstance(x.backend.nplike, TypeTracer) for x in head + tail):
-            head = [
-                x if isinstance(x.backend.nplike, TypeTracer) else x.to_typetracer()
-                for x in head
-            ]
-            tail = [
-                x if isinstance(x.backend.nplike, TypeTracer) else x.to_typetracer()
-                for x in tail
-            ]
-
-        return (head, tail)
+
+        if any(x.backend.nplike.known_data for x in head + tail) and not all(
+            x.backend.nplike.known_data for x in head + tail
+        ):
+            raise RuntimeError
```

Comment:
This can eventually be removed, but we'll leave it in for safety for now. Callers should be enforcing to a common backend.